sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -0,0 +1,214 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.expressions.Literal.FalseLiteral
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.types._
+
+/**
+ * Unwrap casts in binary comparison operations with patterns like following:
+ *
+ * `BinaryComparison(Cast(fromExp, toType), Literal(value, toType))`
+ *   or
+ * `BinaryComparison(Literal(value, toType), Cast(fromExp, toType))`
+ *
+ * This rule optimizes expressions with the above pattern by either replacing the cast with simpler
+ * constructs, or moving the cast from the expression side to the literal side, which enables them
+ * to be optimized away later and pushed down to data sources.
+ *
+ * Currently this only handles cases where `fromType` (of `fromExp`) and `toType` are of integral
+ * types (i.e., byte, short, int and long). The rule checks to see if the literal `value` is
+ * within range `(min, max)`, where `min` and `max` are the minimum and maximum value of
+ * `fromType`, respectively. If this is true then it means we can safely cast `value` to `fromType`
+ * and thus able to move the cast to the literal side.
+ *
+ * If the `value` is not within range `(min, max)`, the rule breaks the scenario into different
+ * cases and try to replace each with simpler constructs.
+ *
+ * if `value > max`, the cases are of following:
+ *  - `cast(exp, ty) > value` ==> if(isnull(exp), null, false)
+ *  - `cast(exp, ty) >= value` ==> if(isnull(exp), null, false)
+ *  - `cast(exp, ty) === value` ==> if(isnull(exp), null, false)
+ *  - `cast(exp, ty) <=> value` ==> false
+ *  - `cast(exp, ty) <= value` ==> if(isnull(exp), null, true)
+ *  - `cast(exp, ty) < value` ==> if(isnull(exp), null, true)
+ *
+ * if `value == max`, the cases are of following:
+ *  - `cast(exp, ty) > value` ==> if(isnull(exp), null, false)
+ *  - `cast(exp, ty) >= value` ==> exp == max
+ *  - `cast(exp, ty) === value` ==> exp == max
+ *  - `cast(exp, ty) <=> value` ==> exp == max
+ *  - `cast(exp, ty) <= value` ==> if(isnull(exp), null, true)
+ *  - `cast(exp, ty) < value` ==> exp =!= max
+ *
+ * Similarly for the cases when `value == min` and `value < min`.
```

Comment:
Does it worth to specifically handle `value == max` and `value == min`? Rules will add cost into planning phase, and we probably not going to run into those two cases often in real world queries. Should we handle them as if it's within the range?

Reply:
It is hard to answer this question without knowing things like how normally user write queries, how much time this rule uses comparing to other rules, and how much time these specific 2 cases use. It seems the he optimization on these two cases can offer significant speed up, and the extra cost is just the pattern matching:

```
else if (maxCmp == 0) {
      exp match {
        case GreaterThan(_, _) =>
          fromExp.falseIfNotNull
        case LessThanOrEqual(_, _) =>
          fromExp.trueIfNotNull
        case LessThan(_, _) =>
          Not(EqualTo(fromExp, Literal(max, fromType)))
        case GreaterThanOrEqual(_, _) | EqualTo(_, _) | EqualNullSafe(_, _) =>
          EqualTo(fromExp, Literal(max, fromType))
        case _ => exp
      }
```

which I think should be pretty fast? comparing to the rule itself which may require walking the whole plan tree.

Reply:
If we don't have `maxCmp == 0` and `minCmp ==0`, it will be optimized to the following. Those can still be pushed down.
```
      val lit = Cast(Literal(value), fromType)
      exp match {
        case GreaterThan(_, _) => GreaterThan(fromExp, lit)
        case GreaterThanOrEqual(_, _) => GreaterThanOrEqual(fromExp, lit)
        case EqualTo(_, _) => EqualTo(fromExp, lit)
        case EqualNullSafe(_, _) => EqualNullSafe(fromExp, lit)
        case LessThan(_, _) => LessThan(fromExp, lit)
        case LessThanOrEqual(_, _) => LessThanOrEqual(fromExp, lit)
        case _ => exp
      }
```

Maybe we can have separate rules that optimize from `GreaterThan(fromExp, lit)` to `fromExp.falseIfNotNull` when lit is the max of lit.dataType. So more plans can take advantage of it.

Reply:
Those filters are not as effective though. We may add separate rules but that may be more expensive I think (we'll need again to check both sides of the comparison, what types the literal is, the min/max, what kind of expression `fromExp` is, and potential edge cases when both sides have different types etc). 

IMO the advantage of doing it here is that we've already done most of the checks and I don't see much extra cost.

Reply:
I was thinking for `value == max`, we can have a separate rule such as

```
 *  - `fromExp > value` where fromExp.type == value.dataType && getRange(fromExp)._2 == value` ==> if(isnull(fromExp), null, false)
...
...
```
which should be valid for other types. Thus, we don't need to handle `maxCmp == 0`, and it opens up other opportunities to optimize rules that don't have `cast`. WDYT?

Reply:
Yes +1 on exploring that. I think we should keep the `maxCmp == 0` case for now though before that is ready, since this is more optimized with not much extra cost (without the `maxCmp == 0` we still have to do pattern matching in the else case).

Reply:
Let's consider it in the followup PR.